### PR TITLE
Add feature to exclude datatypes

### PR DESCRIPTION
**Pull request by jack-gitdev**

### CODE REVIEW

File: warehouse/query-core/src/main/java/datawave/query/tables/ShardQueryLogic.java [hunk 1]
```diff
@@ -746,7 +746,32 @@ protected void loadQueryParameters(ShardQueryConfiguration config, Query setting
...
+                if (dataType.charAt(0) == '!') {
+                    excludeSet.add(StringUtils.substring(dataType, 1));
+                }
+                else {
+                    typeFilter.add(dataType);
+                }
+            }
+
+            if (!excludeSet.isEmpty()) {
+                if (typeFilter.isEmpty()) {
```
Comment (FineAndDandy) [inline warehouse/query-core/src/main/java/datawave/query/tables/ShardQueryLogic.java:761]: throw an error if your ignoring due to the typeFilter being non-empty

**Review by FineAndDandy — APPROVED** [review]

**Review by FineAndDandy — CHANGES_REQUESTED** [review]
can you add a unit test that covers the new functionality?

File: warehouse/query-core/src/main/java/datawave/query/tables/ShardQueryLogic.java [hunk 2]
```diff
@@ -746,7 +746,32 @@ protected void loadQueryParameters(ShardQueryConfiguration config, Query setting
...
+                    MetadataHelper metadataHelper = prepareMetadataHelper(config.getClient(), this.getMetadataTableName(), config.getAuthorizations(), config.isRawTypes());
+
+                    try {
+                        for (Type<?> type : metadataHelper.getAllDatatypes()) {
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/tables/ShardQueryLogic.java]: These are the wrong data types.  I know this is highly confusing, but these data types are the types assigned to fields (e.g. LcNoDiacritics).  Unfortunately we chose the initial "datatype" designation unwisely.  So "datatype" really needs a qualifier of some sort.  Field datatype vs event or ingest datatype ?  IngestDataTypes is used for various method parameters in the MetadataHelper.   Anyway, the method metadataHelper.getDatatypes((Set<String>)null) will get the data types you are looking for.

**Review by ivakegg — CHANGES_REQUESTED** [review]

File: warehouse/query-core/src/main/java/datawave/query/tables/ShardQueryLogic.java [hunk 3]
```diff
@@ -746,7 +746,30 @@ protected void loadQueryParameters(ShardQueryConfiguration config, Query setting
...
+                                    config.isRawTypes());
+
+                    try {
+                        typeFilter.addAll(metadataHelper.getDatatypes(null));
+                    } catch (TableNotFoundException e) {
+                        throw new RuntimeException(e);
+                    }
+                }
+
+                typeFilter.removeAll(excludeSet);
```
Comment (FineAndDandy) [inline warehouse/query-core/src/main/java/datawave/query/tables/ShardQueryLogic.java:810]: consider adding a test that there is at least one dataType left?

**Review by FineAndDandy — CHANGES_REQUESTED** [review]

**Comment by jack-gitdev** [comment]
Need to rework unit tests and add data to testing environment. Current data set does not support testing of necessary edge cases.

**Comment by jack-gitdev** [comment]
Added data to the test data set, adjusted ShardQueryLogicTest to support the expectation of no results, and adjusted unit tests affected by the addition of new data. Removed draft status.

**Review by FineAndDandy — APPROVED** [review]

**Review by ivakegg — APPROVED** [review]
